crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -134,33 +138,159 @@ where
         Ok(())
     }
 
-    pub fn take_args<V: TryFromValue>(
+    pub fn gas_is_taken(&self) -> bool {
+        self.gas.is_none()
+    }
+
+    pub fn take_arg<V: TryFromValue>(
+        &mut self,
+        command_kind: CommandKind<'_>,
+        arg_idx: usize,
+        arg: Argument,
+    ) -> Result<V, ExecutionError> {
+        if matches!(arg, Argument::GasCoin) && !matches!(command_kind, CommandKind::TransferObjects)
+        {
+            panic!("cannot take gas")
+        }
+        if self.arg_is_borrowed(&arg) {
+            panic!("taken borrowed value")
+        }
+        let val_opt = self.borrow_mut(arg)?;
+        if val_opt.is_none() {
+            panic!("taken value")
+        }
+        let val = val_opt.take().unwrap();
+        if let Value::Object(obj) = &val {
+            if let Some(Owner::Shared { .. } | Owner::Immutable) = obj.owner {
+                let error = format!(
+                    "Immutable and shared objects cannot be passed by-value, \
+                                violation found in argument {}",
+                    arg_idx
+                );
+                return Err(ExecutionError::new_with_source(
+                    ExecutionErrorKind::entry_argument_error(
+                        arg_idx as LocalIndex,
+                        EntryArgumentErrorKind::InvalidObjectByValue,
+                    ),
+                    error,
+                ));
+            }
+        }
+        V::try_from_value(val)
+    }
+
+    pub fn borrow_arg_mut<V: TryFromValue>(
+        &mut self,
+        arg_idx: usize,
+        arg: Argument,
+    ) -> Result<V, ExecutionError> {
+        if self.arg_is_borrowed(&arg) {
+            panic!("mutable args can only be used once in a given command")
+        }
+        self.borrowed.insert(arg, /* is_mut */ true);
+        let val_opt = self.borrow_mut(arg)?;
+        if val_opt.is_none() {
+            panic!("taken value")
+        }
+        let val = val_opt.take().unwrap();
+        if let Value::Object(obj) = &val {
+            if let Some(Owner::Immutable) = obj.owner {
+                let error = format!(
+                    "Argument {} is expected to be mutable, immutable object found",
+                    arg_idx
+                );
+                return Err(ExecutionError::new_with_source(
+                    ExecutionErrorKind::entry_argument_error(
+                        arg_idx as LocalIndex,
+                        EntryArgumentErrorKind::InvalidObjectByMuteRef,
+                    ),
+                    error,
+                ));
+            }
+        }
+        V::try_from_value(val)
+    }
+
+    pub fn copy_arg<V: TryFromValue>(
```

Comment:
I did not do that because we will have to check the signature, and I wanted to do that before giving an ability error. For example you call 
```
foo(x: u64)
```
But with a `Coin` as the argument. I want to give a type error, not some sort of "invalid use of copy" error. 

In other words, "copy" here is more about cloning the value from a rust semantics standpoint. Maybe that would be a better name? `clone_arg`?